flink-autoscaler/src/main/java/org/apache/flink/autoscaler/JobVertexScaler.java

```diff
@@ -383,23 +403,68 @@ protected static int scale(
         // Apply min/max parallelism
         newParallelism = Math.min(Math.max(parallelismLowerLimit, newParallelism), upperBound);
 
-        var adjustByMaxParallelism =
-                inputShipStrategies.isEmpty() || inputShipStrategies.contains(HASH);
-        if (!adjustByMaxParallelism) {
+        var adjustByMaxParallelismOrPartitions =
+                numSourcePartitions > 0 || inputShipStrategies.contains(HASH);
+        if (!adjustByMaxParallelismOrPartitions) {
             return newParallelism;
         }
 
-        // When the shuffle type of vertex inputs contains keyBy or vertex is a source, we try to
-        // adjust the parallelism such that it divides the maxParallelism without a remainder
-        // => data is evenly spread across subtasks
-        for (int p = newParallelism; p <= maxParallelism / 2 && p <= upperBound; p++) {
-            if (maxParallelism % p == 0) {
+        final int numKeyGroupsOrPartitions;
+        final int upperBoundForAlignment;
+        if (numSourcePartitions <= 0) {
+            numKeyGroupsOrPartitions = maxParallelism;
+            upperBoundForAlignment =
+                    Math.min(
+                            // Optimize the case where newParallelism <= maxParallelism / 2
+                            newParallelism > maxParallelism / 2
+                                    ? maxParallelism
+                                    : maxParallelism / 2,
+                            upperBound);
+        } else {
+            numKeyGroupsOrPartitions = numSourcePartitions;
+            upperBoundForAlignment = Math.min(numSourcePartitions, upperBound);
+        }
+
+        // When the shuffle type of vertex inputs contains keyBy or vertex is a source,
+        // we try to adjust the parallelism such that it divides
+        // the adjustableMaxParallelism without a remainder => data is evenly spread across subtasks
+        for (int p = newParallelism; p <= upperBoundForAlignment; p++) {
+            if (numKeyGroupsOrPartitions % p == 0) {
                 return p;
             }
         }
 
-        // If parallelism adjustment fails, use originally computed parallelism
-        return newParallelism;
+        // When adjust the parallelism after rounding up cannot be evenly divided by
+        // numKeyGroupsOrPartitions, Try to find the smallest parallelism that can satisfy the
+        // current consumption rate.
+        int p = newParallelism;
+        for (; p > 0; p--) {
+            if (numKeyGroupsOrPartitions / p > numKeyGroupsOrPartitions / newParallelism) {
+                if (numKeyGroupsOrPartitions % p != 0) {
+                    p++;
+                }
+                break;
+            }
+        }
```

Comment:
i found our discussion cannot cover all cases during I review this part in detail.

For example: sourcePartition is 199, and new parallelism is 99. IIUC, the final parallelism is 67(every subtask consume 3 source partitions, except for the last subtask), right?

But 100 as the final parallelism makes sense to me(every subtask consume 2 source partitions, except for the last subtask).

Reply:
> i found our discussion cannot cover all cases during I review this part in detail.
> 
> For example: sourcePartition is 199, and new parallelism is 99. IIUC, the final parallelism is 67(every subtask consume 3 source partitions, except for the last subtask), right?
> 
> But 100 as the final parallelism makes sense to me(every subtask consume 2 source partitions, except for the last subtask).

Well, I actually considered this situation at the very beginning, that is, whether to use the same logic of fetching downwards when fetching upwards, so as to maximize the utilization of resources.

But our ultimate goal is to achieve balanced consumption of tasks? This is the healthiest situation. Often max parallelism or partition will have quite a few divisors. You will rarely set partition num to 1, 3, 7, 199, etc., so when I develop, I first think about balance. If It is due to the limitation of upperbound that we will take a reasonable value downwards, and there will be an event happend.

Reply:
Follow https://github.com/apache/flink-kubernetes-operator/pull/879#discussion_r1764471043 . I found the current logic isn't perfect even if `sourcePartitionNumber` is 200.

Reply:
Good observation, I think there are two opposing ideas in this discussion:

1. Reducing the parallelism as much as possible when we can't evenly balance partitions / key groups
2. Increasing the parallelism such that we can have the most evenly balanced partitions / key groups, even if we cannot do it perfectly (https://github.com/apache/flink-kubernetes-operator/pull/879#discussion_r1764471043)

Both are good ideas. For the sake of stability, I think we probably want to default to doing (2). I could imagine adding an option for (1) but I'm not sure it should be the default mode.